### PR TITLE
Implement docker based command line code executor

### PR DESCRIPTION
**Pull request by jackgerrits**
## Why are these changes needed?

This PR implements Docker based command line code execution in the new code executor framework.

This docker executor also solves the issue where many images are created and polluting the user's Docker installation. It works by running commands in a running container, rather than the old way where it would run each command in a new layer and commit it.

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

File: autogen/coding/docker_commandline_code_executor.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        outputs = []
+        files = []
+        last_exit_code = 0
+        for code_block in code_blocks:
+            lang = code_block.language
+            code = code_block.code
+
+            code_hash = md5(code.encode()).hexdigest()
+
+            # Check if there is a filename comment
```
Comment (ekzhu) [inline autogen/coding/docker_commandline_code_executor.py:166]: Love this. we don't have this for the LocalCommandLineCodeExecutor, should we reuse this code?

Comment (jackgerrits) [reply autogen/coding/docker_commandline_code_executor.py:166]: Yeah I think we should do it! I can do it in a separate PR

File: autogen/coding/docker_commandline_code_executor.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    The executor sanitizes command line commands using regular expression
+    match against a list of dangerous commands in order to prevent
+    self-destructive commands from being executed which may potentially
+    affect the host environment.
+    Currently, the executor only supports Python and shell scripts.
+    For Python code, use the language "python" for the code block.
+    For shell scripts, use the language "bash", "shell", or "sh" for the code
+    block.
+
+    Args:
```
Comment (ekzhu) [inline autogen/coding/docker_commandline_code_executor.py:57]: Doc for each of the constructor arguments.

File: autogen/coding/docker_commandline_code_executor.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+            a default working directory will be used. The default working
+            directory is the current directory ".".
+    """
+
+    def __init__(
+        self,
+        image: str = "python:3-slim",
+        container_name: Optional[str] = None,
+        timeout: int = 60,
+        work_dir: Union[Path, str] = Path("."),
```
Comment (ekzhu) [inline autogen/coding/docker_commandline_code_executor.py:69]: In `LocalCommandLineExecutor` the work_dir argument is string only. Let's make the interface consistent in the next PR.

File: website/docs/topics/code-execution/cli-code-executor.ipynb
Comment (ekzhu) [inline website/docs/topics/code-execution/cli-code-executor.ipynb:1]: In the Local example, we will remove the str() wround work_dir once the LocalCommandLineExecutor's accepts Path type.

**Review by ekzhu — APPROVED** [review]

**Review by ekzhu — CHANGES_REQUESTED** [review]
Some docs on constructor is needed. Some note for the next PR.

**Review by ekzhu — APPROVED** [review]

**Pull request merged**